File: python/todo_cli/todo_cli/commands.py
from datetime import datetime

from rich.console import Console
from rich.table import Table

from .models import Priority, Todo, TodoStore
from .storage import load, save

console = Console()


def add(text: str, priority: Priority, tags: list[str]) -> None:
    store = load()
    todo = Todo(id=store.next_id, text=text, priority=priority, tags=tags)
    store.todos.append(todo)
    store.next_id += 1
    save(store)
    console.print(f"[green]Added #{todo.id}:[/green] {text}")


def list_todos(show_done: bool, tag: str | None, priority: Priority | None) -> None:
    store = load()
    todos = store.todos

    if not show_done:
        todos = [t for t in todos if not t.done]
    if tag:
        todos = [t for t in todos if tag in t.tags]
    if priority:
        todos = [t for t in todos if t.priority == priority]

    if not todos:
        console.print("[dim]No todos found.[/dim]")
        return

    table = Table(title="Todos")
    table.add_column("ID", style="cyan", width=4)
    table.add_column("Status", width=6)
    table.add_column("Pri", width=6)
    table.add_column("Text")
    table.add_column("Tags", style="dim")
    table.add_column("Created", style="dim")

    priority_colors = {Priority.HIGH: "red", Priority.MEDIUM: "yellow", Priority.LOW: "green"}

    for t in todos:
        status = "[green]done[/green]" if t.done else "[dim]open[/dim]"
        color = priority_colors[t.priority]
        pri = f"[{color}]{t.priority.value}[/{color}]"
        tags = ", ".join(t.tags) if t.tags else ""
        created = t.created_at[:10]
        table.add_row(str(t.id), status, pri, t.text, tags, created)

    console.print(table)


def done(todo_id: int) -> None:
    store = load()
    for t in store.todos:
        if t.id == todo_id:
            assert not t.done, f"Todo #{todo_id} is already done"
            t.done = True
            t.completed_at = datetime.now().isoformat()
            save(store)
            console.print(f"[green]Completed #{todo_id}:[/green] {t.text}")
            return
    console.print(f"[red]Not found: #{todo_id}[/red]")


def remove(todo_id: int) -> None:
    store = load()
    before = len(store.todos)
    store.todos = [t for t in store.todos if t.id != todo_id]
    assert len(store.todos) < before, f"Not found: #{todo_id}"
    save(store)
    console.print(f"[red]Removed #{todo_id}[/red]")


def edit(todo_id: int, text: str | None, priority: Priority | None, tags: list[str] | None) -> None:
    store = load()
    for t in store.todos:
        if t.id == todo_id:
            if text is not None:
                t.text = text
            if priority is not None:
                t.priority = priority
            if tags is not None:
                t.tags = tags
            save(store)
            console.print(f"[yellow]Updated #{todo_id}[/yellow]")
            return
    console.print(f"[red]Not found: #{todo_id}[/red]")


def stats() -> None:
    store = load()
    total = len(store.todos)
    completed = sum(1 for t in store.todos if t.done)
    pending = total - completed

    by_priority = {}
    for t in store.todos:
        if not t.done:
            by_priority[t.priority] = by_priority.get(t.priority, 0) + 1

    console.print(f"\n[bold]Stats[/bold]")
    console.print(f"  Total: {total}  |  Pending: {pending}  |  Done: {completed}")
    if by_priority:
        parts = [f"{p.value}: {c}" for p, c in sorted(by_priority.items())]
        console.print(f"  By priority: {', '.join(parts)}")
    console.print()

File: elixir/todo_cli/lib/todo_cli/commands.ex
defmodule TodoCli.Commands do
  alias TodoCli.{Models, Storage}

  def add(text, priority, tags) do
    {todos, next_id} = Storage.load()
    now = DateTime.utc_now() |> DateTime.to_iso8601()
    todo = %Models{id: next_id, text: text, priority: priority, tags: tags, created_at: now}
    Storage.save(todos ++ [todo], next_id + 1)
    IO.puts(IO.ANSI.green() <> "Added ##{next_id}: " <> IO.ANSI.reset() <> text)
  end

  def list_todos(show_done, tag, priority) do
    {todos, _} = Storage.load()
    todos = if show_done, do: todos, else: Enum.filter(todos, &(!&1.done))
    todos = if tag, do: Enum.filter(todos, &(tag in &1.tags)), else: todos
    todos = if priority, do: Enum.filter(todos, &(&1.priority == priority)), else: todos

    if todos == [] do
      IO.puts(IO.ANSI.faint() <> "No todos found." <> IO.ANSI.reset())
    else
      header = format_row("ID", "Status", "Pri", "Text", "Tags", "Created")
      IO.puts(IO.ANSI.bright() <> header <> IO.ANSI.reset())
      Enum.each(todos, &print_todo/1)
    end
  end

  defp format_row(id, status, pri, text, tags, created) do
    String.pad_trailing(id, 6) <>
      String.pad_trailing(status, 8) <>
      String.pad_trailing(pri, 8) <>
      String.pad_trailing(text, 30) <>
      String.pad_trailing(tags, 16) <>
      created
  end

  defp priority_color("high"), do: IO.ANSI.red()
  defp priority_color("medium"), do: IO.ANSI.yellow()
  defp priority_color("low"), do: IO.ANSI.green()

  defp print_todo(t) do
    status = if t.done, do: IO.ANSI.green() <> "done" <> IO.ANSI.reset(), else: IO.ANSI.faint() <> "open" <> IO.ANSI.reset()
    pri = priority_color(t.priority) <> t.priority <> IO.ANSI.reset()
    tags = Enum.join(t.tags, ", ")
    created = String.slice(t.created_at || "", 0, 10)
    IO.puts(format_row("#{t.id}", status, pri, t.text, tags, created))
  end

  def done(todo_id) do
    {todos, next_id} = Storage.load()
    {todo, rest} = pop_by_id(todos, todo_id)
    assert todo != nil, "Not found: ##{todo_id}"
    assert !todo.done, "Todo ##{todo_id} is already done"
    now = DateTime.utc_now() |> DateTime.to_iso8601()
    updated = %{todo | done: true, completed_at: now}
    Storage.save(reinsert(rest, updated, todos), next_id)
    IO.puts(IO.ANSI.green() <> "Completed ##{todo_id}: " <> IO.ANSI.reset() <> todo.text)
  end

  def remove(todo_id) do
    {todos, next_id} = Storage.load()
    filtered = Enum.reject(todos, &(&1.id == todo_id))
    assert length(filtered) < length(todos), "Not found: ##{todo_id}"
    Storage.save(filtered, next_id)
    IO.puts(IO.ANSI.red() <> "Removed ##{todo_id}" <> IO.ANSI.reset())
  end

  def edit(todo_id, text, priority, tags) do
    {todos, next_id} = Storage.load()
    {todo, rest} = pop_by_id(todos, todo_id)
    assert todo != nil, "Not found: ##{todo_id}"
    updated = todo
    updated = if text, do: %{updated | text: text}, else: updated
    updated = if priority, do: %{updated | priority: priority}, else: updated
    updated = if tags, do: %{updated | tags: tags}, else: updated
    Storage.save(reinsert(rest, updated, todos), next_id)
    IO.puts(IO.ANSI.yellow() <> "Updated ##{todo_id}" <> IO.ANSI.reset())
  end

  def stats do
    {todos, _} = Storage.load()
    total = length(todos)
    completed = Enum.count(todos, & &1.done)
    pending = total - completed

    by_priority =
      todos
      |> Enum.reject(& &1.done)
      |> Enum.frequencies_by(& &1.priority)
      |> Enum.sort()

    IO.puts("\n#{IO.ANSI.bright()}Stats#{IO.ANSI.reset()}")
    IO.puts("  Total: #{total}  |  Pending: #{pending}  |  Done: #{completed}")

    if by_priority != [] do
      parts = Enum.map_join(by_priority, ", ", fn {p, c} -> "#{p}: #{c}" end)
      IO.puts("  By priority: #{parts}")
    end

    IO.puts("")
  end

  defp pop_by_id(todos, id) do
    case Enum.find(todos, &(&1.id == id)) do
      nil -> {nil, todos}
      todo -> {todo, Enum.reject(todos, &(&1.id == id))}
    end
  end

  defp reinsert(rest, updated, original) do
    idx = Enum.find_index(original, &(&1.id == updated.id))
    List.insert_at(rest, idx, updated)
  end

  defp assert(true, _), do: :ok
  defp assert(false, msg), do: raise(msg)
end
